## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

data_df = pd.read_csv('Resources/client_dataset.csv')
data_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
data_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
data_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data
data_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# What three item categories had the most entries?
category_counts_df = data_df['category'].value_counts()  # Count occurrences of each category
top_categories_df = category_counts_df.head(3) #Get the top 3 categories
print(top_categories_df)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [6]:
# For the category with the most entries, which subcategory had the most entries?
most_entries_df = data_df['category'].value_counts()
most_entries_category_df = most_entries_df.idxmax()
most_entries_subcategory_df = data_df[data_df['category'] == most_entries_category_df]['subcategory'].value_counts()
most_entries_subcategory_df.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
top_5_clients = data_df['client_id'].value_counts()  
top_5_clients.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_5_clients_list = top_5_clients.head(5).index.tolist()
print(top_5_clients.head(5))

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [9]:
# How many total units (the qty column) did the client with the most entries order ordered
total_units = data_df[data_df['client_id'] == top_5_clients_list[0]]['qty'].sum()
total_units

np.int64(64313)

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
sub_data_df = data_df[['unit_price', 'qty']].copy()
data_df['line_subtotal'] = sub_data_df['unit_price'] * sub_data_df['qty']
data_df[['unit_price', 'qty', 'line_subtotal']].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 
# per pound for items 50 pounds or under.
total_weight_df = data_df[['unit_weight', 'qty']].copy()
data_df['total_weight'] = total_weight_df['unit_weight'] * total_weight_df['qty']

def shipping_price(row):
    if row['total_weight'] > 50:
        return row['total_weight'] * 7
    else:
        return row['total_weight'] * 10
    
data_df['shipping_price'] = data_df.apply(shipping_price, axis=1)
data_df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
line_price = data_df[['line_subtotal', 'shipping_price']]
data_df['line_price'] = data_df['line_subtotal'] + data_df['shipping_price'] + (data_df['line_subtotal'] + data_df['shipping_price']) * 0.0925
data_df[['line_subtotal', 'shipping_price', 'line_price']].round(2).head(3)


line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

line_cost = data_df[['unit_cost', 'qty', 'shipping_price']].copy()
data_df['line_cost'] = line_cost['unit_cost'] * line_cost['qty'] + line_cost['shipping_price']
data_df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price     line_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price

line_profit = data_df[['line_subtotal', 'total_weight', 'shipping_price']]

data_df['line_price'] = data_df['line_subtotal'] + data_df['shipping_price']\
     + (data_df['line_subtotal'] + data_df['shipping_price']) * 0.0925

data_df['line_profit'] = data_df['line_price'] - data_df['line_cost']
data_df.round(2).head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above

order_id_to_find = 2742071
total_ID = {order_id_to_find: data_df[data_df['order_id'] == order_id_to_find]['line_price'].sum()}
for key, value in total_ID.items():
    print(f"Order ID {key} had a total price of {round(value, 2)}")

order_id_to_find = 2173913
total_ID = {order_id_to_find: data_df[data_df['order_id'] == order_id_to_find]['line_price'].sum()}
for key, value in total_ID.items():
    print(f"Order ID {key} had a total price of {round(value, 2)}")

order_id_to_find = 6128929
total_ID = {order_id_to_find: data_df[data_df['order_id'] == order_id_to_find]['line_price'].sum()}
for key, value in total_ID.items():
    print(f"Order ID {key} had a total price of {round(value, 2)}")


Order ID 2742071 had a total price of 152811.9
Order ID 2173913 had a total price of 162388.72
Order ID 6128929 had a total price of 923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? 
# Check your work from Part 1 for client ids.
top_5_clients_df = data_df['client_id'].value_counts()

top_5_clients_total = data_df[data_df['client_id'].isin(top_5_clients_list)].groupby('client_id')['line_price'].sum()

for client_id, total_spent in top_5_clients_total.items():
    print(f"Client {client_id}: ${total_spent:.2f}")


Client 24741: $82268891.98
Client 33615: $8377308.57
Client 38378: $12906550.88
Client 46820: $9743794.32
Client 66037: $10259514.80


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

top_5_clients_total = data_df[data_df['client_id'].isin(top_5_clients_list)].groupby('client_id')\
    [['qty','shipping_price','line_price', 'line_cost', 'line_profit']].sum().round(2)

top_5_clients_total.head(5)


qty  shipping_price   line_price    line_cost  line_profit
client_id                                                               
24741      239862      5126448.37  82268891.98  45688899.71  36579992.27
33615       64313      1828984.89   8377308.57   6175313.91   2201994.66
38378       73667      3429455.40  12906550.88   9634720.98   3271829.90
46820       75768      1601448.84   9743794.32   7007191.64   2736602.68
66037       43018      1395151.85  10259514.80   7004482.98   3255031.82

In [18]:
# Format the data and rename the columns to names suitable for presentation.
data_df.rename(columns={
    'first': 'First Name',
    'last': 'Last Name',
    'job': 'Job',
    'phone': 'Phone',
    'email': 'Email',
    'client_id': 'Client ID',
    'order_id': 'Order ID',
    'order_date': 'Order Date',
    'order_week': 'Order Week',
    'order_year': 'Order Year',
    'item_id': 'Item',
    'category': 'Category',
    'subcategory': 'Subcategory',
    'unit_price': 'Unit Price',
    'unit_cost': 'Unit Cost',
    'unit_weight': 'Unit Weight',
    'qty': 'Units',
    'line_number': 'Line Number',
    'line_subtotal': 'Line Subtotal',
    'total_weight': 'Total Weight',
    'shipping_price': 'Shipping Price',
    'line_price': 'Total Price',
    'line_cost': 'Total Cost',
    'line_profit': 'Total Profit'
}, inplace=True)

data_df.head(5) 


First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ... Unit Cost Unit Weight Units  Line Number  \
0          17        2023  ...    762.71        7.50   105            1   
1          20        2023  ...     15.09        1.49    21            0   
2           4        2023  ...      7.86        1.68    39            6   
3          17        2023  ...     24.85        1.23    29            3   
4          21        2023  ...    108.17       46.43    20            1   

   Line Subtotal  Total Weight  Shipping Price    Total Price  Total Cost  \
0      115164.00        787.50         5512.50  131839.076250    85597.05   
1         523.95         31.29          312.90     914.258625      629.79   
2         527.28         65.52          458.64    1077.117600      765.18   
3        1056.18         35.67          356.70    1543.571400     1077.35   
4        3902.00        928.60         6500.20   11364.403500     8663.60   

   Total Profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

In [19]:
# Define the money columns. 
money_columns = ['Shipping Price', 'Total Price', 'Total Cost',
                 'Total Profit']
money_columns

['Shipping Price', 'Total Price', 'Total Cost', 'Total Profit']

In [20]:
### Define a function that converts a dollar amount to millions.
def convert_to_millions(money_columns):
    return money_columns / 1000000


In [21]:
# Apply the currency_format_millions function to only the money columns.
# Rename the columns to reflect the change in the money format. 
data_df.rename(columns={
    'Shipping Price': 'Shipping Price (millions)', 
    'Total Price': 'Total Price (millions)', 
    'Total Cost': 'Total Cost (millions)', 
    'Total Profit': 'Total Profit (millions)'
}, inplace=True)

data_df['Shipping Price (millions)'] = convert_to_millions(data_df['Shipping Price (millions)'])
data_df['Total Price (millions)'] = convert_to_millions(data_df['Total Price (millions)'])
data_df['Total Cost (millions)'] = convert_to_millions(data_df['Total Cost (millions)'])
data_df['Total Profit (millions)'] = convert_to_millions(data_df['Total Profit (millions)'])

top_5_clients = data_df[data_df['Client ID'].isin(top_5_clients_list)][['Client ID', 'Units', 'Shipping Price (millions)', 'Total Cost (millions)', 'Total Price (millions)', 'Total Profit (millions)']]

top_5_clients_total = top_5_clients.groupby('Client ID').sum()
top_5_clients_total

Units  Shipping Price (millions)  Total Cost (millions)  \
Client ID                                                             
24741      239862                   5.126448              45.688900   
33615       64313                   1.828985               6.175314   
38378       73667                   3.429455               9.634721   
46820       75768                   1.601449               7.007192   
66037       43018                   1.395152               7.004483   

           Total Price (millions)  Total Profit (millions)  
Client ID                                                   
24741                   82.268892                36.579992  
33615                    8.377309                 2.201995  
38378                   12.906551                 3.271830  
46820                    9.743794                 2.736603  
66037                   10.259515                 3.255032

In [22]:
### Sort the updated data by "Total Profit (millions)" from highest to lowest and 
### assign the sort to a new DataFrame.
top_5_clients_total_sorted = top_5_clients.groupby('Client ID').sum()
top_5_clients_total_sorted_df = top_5_clients_total_sorted.sort_values(by="Total Profit (millions)", ascending=False)
top_5_clients_total_sorted_df


Units  Shipping Price (millions)  Total Cost (millions)  \
Client ID                                                             
24741      239862                   5.126448              45.688900   
38378       73667                   3.429455               9.634721   
66037       43018                   1.395152               7.004483   
46820       75768                   1.601449               7.007192   
33615       64313                   1.828985               6.175314   

           Total Price (millions)  Total Profit (millions)  
Client ID                                                   
24741                   82.268892                36.579992  
38378                   12.906551                 3.271830  
66037                   10.259515                 3.255032  
46820                    9.743794                 2.736603  
33615                    8.377309                 2.201995

In [ ]:
# The analysis reveals that the top 5 clients, based on the quantity of units purchased,
# have significant variations in their total spending and profits. 
# Client ID 24741 stands out with the highest total profit of $36.58 million, 
# while Client ID 33615 has the lowest total profit among the top 5, at $2.20 million. 
# This indicates a substantial difference in profitability among the top clients.